argon2/src/instance.rs

```diff
@@ -115,10 +122,54 @@ impl<'a> Instance<'a> {
         Ok(instance)
     }
 
+    /// Create multiple mutable references for the current instance, one for every lane
+    #[cfg(feature = "parallel")]
+    #[allow(unsafe_code)]
+    unsafe fn mut_self_refs(&mut self) -> Vec<&'static mut Instance<'static>> {
+        let lanes = self.lanes;
+        // This transmute can be skipped when a scoped threadpool is used (or when `spawn_unchecked()` gets stabilised)
+        let this = mem::transmute::<_, &mut Instance<'static>>(self);
+        let this: *mut Instance<'static> = this;
+
+        // Dereference the raw pointer multiple times to create multiple mutable references
+        (0..lanes).map(|_| &mut *(this)).collect()
```

Comment:
Under the unsafe code guidelines we've been considering, this would definitely be considered UB (or, at least, it would be if you go on to use these `&mut` later). Each time you create an `&mut`, you are effectively "stealing" the access permission to the memory from the previous one, and only one `&mut` can have that at a time.

Reply:
What would be valid is if you were doing something like `&mut (*this + i)` -- i.e., creating `&mut references to disjoint areas in memory.

Reply:
It would *also* be ok to return a `Vec<*mut Instance>` and then ensure that the access you do are disjoint -- but that is pushing some unsafety onto the users of the raw pointers later (which is correct and valid).

Reply:
Consider, with the code as is:

You could, in safe code, create threads that access the same fields from these `&mut Instance` pointers, and each of those threads would (rightly) think that it has unique access.

Reply:
> It would also be ok to return a Vec<*mut Instance> and then ensure that the access you do are disjoint -- but that is pushing some unsafety onto the users of the raw pointers later (which is correct and valid)

So wrapping the raw pointer to the instance inside an `AtomicPtr` could work as a band aid fix?